File: packages/shared/src/client-workouts.ts
import { z } from 'zod';

const name = z.string().trim().min(1).max(200);
const optInt = z.number().int().positive().nullish();
const optNum = z.number().positive().nullish();

export const workoutStatusSchema = z.enum(['draft', 'active', 'completed', 'skipped']);
export type WorkoutStatus = z.infer<typeof workoutStatusSchema>;

// --- Создание плана тренировки ---

export const plannedSetSchema = z.object({
  plannedReps: optInt,
  plannedWeightKg: optNum,
  plannedTimeSec: optInt,
  plannedRestSec: z.number().int().min(0).max(3600).nullish(),
});
export type PlannedSet = z.infer<typeof plannedSetSchema>;

export const workoutExercisePlanSchema = z.object({
  exerciseId: z.string(),
  sets: z.array(plannedSetSchema).min(1),
});
export type WorkoutExercisePlan = z.infer<typeof workoutExercisePlanSchema>;

export const createWorkoutRequestSchema = z.object({
  name,
  sourceTemplateId: z.string().nullish(),
  // Допускаем пустой список: клиент создаёт пустую тренировку и наполняет её
  // упражнениями уже на странице проведения (как тренер в ActiveWorkout).
  exercises: z.array(workoutExercisePlanSchema),
  // Тренер формирует историческую запись (постфактум): не уведомлять клиента и
  // не учитывать в балансе пакета/календаре. Для клиентских тренировок игнорируется.
  excludedFromBalance: z.boolean().optional(),
});
export type CreateWorkoutRequest = z.infer<typeof createWorkoutRequestSchema>;

// Тренер фиксирует уже проведённую тренировку в истории клиента указанной датой.
export const addWorkoutToHistoryRequestSchema = z.object({
  date: z.string().regex(/^\d{4}-\d{2}-\d{2}$/, 'Ожидается дата YYYY-MM-DD'),
});
export type AddWorkoutToHistoryRequest = z.infer<typeof addWorkoutToHistoryRequestSchema>;

// --- Редактирование набора упражнений тренировки ---

// Добавление упражнения = тот же план одной позиции (exerciseId + plannedSets).
export const addWorkoutExerciseRequestSchema = workoutExercisePlanSchema;
export type AddWorkoutExerciseRequest = z.infer<typeof addWorkoutExerciseRequestSchema>;

// Добавление ОДНОГО подхода к упражнению (плановые параметры, все опциональны).
export const addWorkoutSetRequestSchema = plannedSetSchema;
export type AddWorkoutSetRequest = z.infer<typeof addWorkoutSetRequestSchema>;

// order = массив текущих position в новом порядке (перестановка существующих позиций).
export const reorderWorkoutExercisesRequestSchema = z.object({
  order: z.array(z.number().int().min(0)).min(1),
});
export type ReorderWorkoutExercisesRequest = z.infer<typeof reorderWorkoutExercisesRequestSchema>;

// --- Фиксация факта по подходу ---

export const updateSetRequestSchema = z.object({
  plannedReps: z.number().int().positive().nullish(),
  plannedWeightKg: z.number().positive().nullish(),
  plannedTimeSec: z.number().int().positive().nullish(),
  plannedRestSec: z.number().int().min(0).max(3600).nullish(),
  actualReps: z.number().int().nullish(),
  actualWeightKg: optNum,
  actualTimeSec: z.number().int().nullish(),
  done: z.boolean().optional(),
});
export type UpdateSetRequest = z.infer<typeof updateSetRequestSchema>;

// --- Завершение тренировки ---

export const completeWorkoutRequestSchema = z.object({
  durationSec: optInt,
  trainerNote: z.string().trim().max(2000).nullish(),
  rpe: z.number().int().min(1).max(10).nullish(),
  // Смещение таймзоны устройства тренера от UTC в минутах (как Dart
  // timeZoneOffset.inMinutes, +180 для МСК). По нему сервер проставляет ЛОКАЛЬНОЕ
  // время авто-созданного занятия (иначе оно бы шло по времени сервера/UTC).
  tzOffsetMinutes: z.number().int().nullish(),
});
export type CompleteWorkoutRequest = z.infer<typeof completeWorkoutRequestSchema>;

// --- Ответы ---

export const workoutSetResponseSchema = z.object({
  setIndex: z.number(),
  plannedReps: z.number().nullable(),
  plannedWeightKg: z.number().nullable(),
  plannedTimeSec: z.number().nullable(),
  plannedRestSec: z.number().nullable(),
  actualReps: z.number().nullable(),
  actualWeightKg: z.number().nullable(),
  actualTimeSec: z.number().nullable(),
  done: z.boolean(),
});
export type WorkoutSetResponse = z.infer<typeof workoutSetResponseSchema>;

export const workoutExerciseResponseSchema = z.object({
  position: z.number(),
  exerciseId: z.string(),
  exerciseName: z.string(),
  sets: z.array(workoutSetResponseSchema),
});
export type WorkoutExerciseResponse = z.infer<typeof workoutExerciseResponseSchema>;

export const workoutResponseSchema = z.object({
  id: z.string(),
  clientId: z.string(),
  name: z.string(),
  status: workoutStatusSchema,
  startedAt: z.string().nullable(),
  completedAt: z.string().nullable(),
  durationSec: z.number().nullable(),
  trainerNote: z.string().nullable(),
  rpe: z.number().nullable(),
  createdByClient: z.boolean(),
  // Историческая запись тренера: не уменьшает баланс пакета, нет в календаре.
  excludedFromBalance: z.boolean(),
  exercises: z.array(workoutExerciseResponseSchema),
});
export type WorkoutResponse = z.infer<typeof workoutResponseSchema>;

export const workoutListResponseSchema = z.object({
  workouts: z.array(workoutResponseSchema),
});
export type WorkoutListResponse = z.infer<typeof workoutListResponseSchema>;

File: mobile/apps/client/lib/api/client_workouts.dart
import 'package:core/core.dart';
import 'package:flutter_riverpod/flutter_riverpod.dart';

/// Статус тренировки (зеркало workoutStatusSchema).
enum WorkoutStatus { draft, active, completed, skipped, unknown }

WorkoutStatus _statusFrom(String? s) => switch (s) {
      'draft' => WorkoutStatus.draft,
      'active' => WorkoutStatus.active,
      'completed' => WorkoutStatus.completed,
      'skipped' => WorkoutStatus.skipped,
      _ => WorkoutStatus.unknown,
    };

/// Подход упражнения (план + факт).
class WorkoutSet {
  WorkoutSet({
    required this.setIndex,
    required this.plannedReps,
    required this.plannedWeightKg,
    required this.plannedTimeSec,
    required this.actualReps,
    required this.actualWeightKg,
    required this.actualTimeSec,
    required this.done,
  });

  final int setIndex;
  final num? plannedReps;
  final num? plannedWeightKg;
  final num? plannedTimeSec;
  final num? actualReps;
  final num? actualWeightKg;
  final num? actualTimeSec;
  final bool done;

  factory WorkoutSet.fromJson(Map<String, dynamic> j) => WorkoutSet(
        setIndex: (j['setIndex'] as num?)?.toInt() ?? 0,
        plannedReps: j['plannedReps'] as num?,
        plannedWeightKg: j['plannedWeightKg'] as num?,
        plannedTimeSec: j['plannedTimeSec'] as num?,
        actualReps: j['actualReps'] as num?,
        actualWeightKg: j['actualWeightKg'] as num?,
        actualTimeSec: j['actualTimeSec'] as num?,
        done: j['done'] as bool? ?? false,
      );
}

/// Упражнение тренировки с набором подходов.
class WorkoutExercise {
  WorkoutExercise({required this.position, required this.name, required this.sets});
  final int position;
  final String name;
  final List<WorkoutSet> sets;

  factory WorkoutExercise.fromJson(Map<String, dynamic> j) => WorkoutExercise(
        position: (j['position'] as num?)?.toInt() ?? 0,
        name: j['exerciseName'] as String? ?? 'Упражнение',
        sets: ((j['sets'] as List<dynamic>?) ?? <dynamic>[])
            .cast<Map<String, dynamic>>()
            .map(WorkoutSet.fromJson)
            .toList(),
      );
}

/// Тренировка (зеркало workoutResponseSchema).
class Workout {
  Workout({
    required this.id,
    required this.name,
    required this.status,
    required this.completedAt,
    required this.durationSec,
    required this.rpe,
    required this.trainerNote,
    required this.createdByClient,
    required this.exercises,
  });

  final String id;
  final String name;
  final WorkoutStatus status;
  final DateTime? completedAt;
  final int? durationSec;
  final num? rpe;
  final String? trainerNote;
  final bool createdByClient;
  final List<WorkoutExercise> exercises;

  factory Workout.fromJson(Map<String, dynamic> j) {
    final String? ca = j['completedAt'] as String?;
    return Workout(
      id: j['id'] as String? ?? '',
      name: (j['name'] as String? ?? '').trim().isNotEmpty ? j['name'] as String : 'Тренировка',
      status: _statusFrom(j['status'] as String?),
      completedAt: ca != null ? DateTime.tryParse(ca)?.toLocal() : null,
      durationSec: (j['durationSec'] as num?)?.toInt(),
      rpe: j['rpe'] as num?,
      trainerNote: j['trainerNote'] as String?,
      createdByClient: j['createdByClient'] as bool? ?? false,
      exercises: ((j['exercises'] as List<dynamic>?) ?? <dynamic>[])
          .cast<Map<String, dynamic>>()
          .map(WorkoutExercise.fromJson)
          .toList(),
    );
  }

  int get totalSets => exercises.fold<int>(0, (int a, WorkoutExercise e) => a + e.sets.length);
}

/// Доступ к тренировкам клиента: список (назначенные + история).
class ClientWorkoutsApi {
  ClientWorkoutsApi(this._ref);
  final Ref _ref;
  ApiClient get _api => _ref.read(apiClientProvider);

  Future<List<Workout>> load() async {
    final Map<String, dynamic> r = await _api.getJson('/api/client/workouts');
    final List<Workout> list = ((r['workouts'] as List<dynamic>?) ?? <dynamic>[])
        .cast<Map<String, dynamic>>()
        .map(Workout.fromJson)
        .toList();
    // Свежие завершённые — выше; черновики — по имени.
    list.sort((Workout a, Workout b) =>
        (b.completedAt ?? DateTime.fromMillisecondsSinceEpoch(0))
            .compareTo(a.completedAt ?? DateTime.fromMillisecondsSinceEpoch(0)));
    return list;
  }
}

final Provider<ClientWorkoutsApi> clientWorkoutsApiProvider =
    Provider<ClientWorkoutsApi>((ref) => ClientWorkoutsApi(ref));

final FutureProvider<List<Workout>> clientWorkoutsProvider =
    FutureProvider<List<Workout>>((ref) => ref.read(clientWorkoutsApiProvider).load());
